File: mediapipe/tasks/cc/vision/image_segmenter/image_segmenter.cc
/* Copyright 2022 The MediaPipe Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "mediapipe/tasks/cc/vision/image_segmenter/image_segmenter.h"

#include <optional>
#include <utility>

#include "absl/strings/str_format.h"
#include "mediapipe/framework/api2/builder.h"
#include "mediapipe/framework/calculator.pb.h"
#include "mediapipe/framework/formats/image.h"
#include "mediapipe/framework/formats/rect.pb.h"
#include "mediapipe/framework/port/status_macros.h"
#include "mediapipe/tasks/cc/core/utils.h"
#include "mediapipe/tasks/cc/vision/core/image_processing_options.h"
#include "mediapipe/tasks/cc/vision/core/running_mode.h"
#include "mediapipe/tasks/cc/vision/core/vision_task_api_factory.h"
#include "mediapipe/tasks/cc/vision/image_segmenter/calculators/tensors_to_segmentation_calculator.pb.h"
#include "mediapipe/tasks/cc/vision/image_segmenter/proto/image_segmenter_graph_options.pb.h"
#include "mediapipe/tasks/cc/vision/image_segmenter/proto/segmenter_options.pb.h"
#include "mediapipe/util/label_map.pb.h"

namespace mediapipe {
namespace tasks {
namespace vision {
namespace image_segmenter {
namespace {

constexpr char kConfidenceMasksTag[] = "CONFIDENCE_MASKS";
constexpr char kConfidenceMasksStreamName[] = "confidence_masks";
constexpr char kCategoryMaskTag[] = "CATEGORY_MASK";
constexpr char kCategoryMaskStreamName[] = "category_mask";
constexpr char kOutputSizeTag[] = "OUTPUT_SIZE";
constexpr char kOutputSizeStreamName[] = "output_size";
constexpr char kImageInStreamName[] = "image_in";
constexpr char kImageOutStreamName[] = "image_out";
constexpr char kImageTag[] = "IMAGE";
constexpr char kNormRectStreamName[] = "norm_rect_in";
constexpr char kNormRectTag[] = "NORM_RECT";
constexpr char kQualityScoresStreamName[] = "quality_scores";
constexpr char kQualityScoresTag[] = "QUALITY_SCORES";
constexpr char kSubgraphTypeName[] =
    "mediapipe.tasks.vision.image_segmenter.ImageSegmenterGraph";
constexpr int kMicroSecondsPerMilliSecond = 1000;

using ::mediapipe::CalculatorGraphConfig;
using ::mediapipe::Image;
using ::mediapipe::NormalizedRect;
using ImageSegmenterGraphOptionsProto = ::mediapipe::tasks::vision::
    image_segmenter::proto::ImageSegmenterGraphOptions;

// Creates a MediaPipe graph config that only contains a single subgraph node of
// "mediapipe.tasks.vision.image_segmenter.ImageSegmenterGraph".
CalculatorGraphConfig CreateGraphConfig(
    std::unique_ptr<ImageSegmenterGraphOptionsProto> options,
    bool output_confidence_masks, bool output_category_mask,
    bool enable_flow_limiting) {
  api2::builder::Graph graph;
  auto& task_subgraph = graph.AddNode(kSubgraphTypeName);
  task_subgraph.GetOptions<ImageSegmenterGraphOptionsProto>().Swap(
      options.get());
  graph.In(kImageTag).SetName(kImageInStreamName);
  graph.In(kNormRectTag).SetName(kNormRectStreamName);
  graph.In(kOutputSizeTag).SetName(kOutputSizeStreamName);
  if (output_confidence_masks) {
    task_subgraph.Out(kConfidenceMasksTag)
            .SetName(kConfidenceMasksStreamName) >>
        graph.Out(kConfidenceMasksTag);
  }
  if (output_category_mask) {
    task_subgraph.Out(kCategoryMaskTag).SetName(kCategoryMaskStreamName) >>
        graph.Out(kCategoryMaskTag);
  }
  task_subgraph.Out(kQualityScoresTag).SetName(kQualityScoresStreamName) >>
      graph.Out(kQualityScoresTag);
  task_subgraph.Out(kImageTag).SetName(kImageOutStreamName) >>
      graph.Out(kImageTag);
  if (enable_flow_limiting) {
    return tasks::core::AddFlowLimiterCalculator(
        graph, task_subgraph, {kImageTag, kNormRectTag, kOutputSizeTag},
        kConfidenceMasksTag);
  }
  graph.In(kImageTag) >> task_subgraph.In(kImageTag);
  graph.In(kNormRectTag) >> task_subgraph.In(kNormRectTag);
  graph.In(kOutputSizeTag) >> task_subgraph.In(kOutputSizeTag);
  return graph.GetConfig();
}

// Converts the user-facing ImageSegmenterOptions struct to the internal
// ImageSegmenterOptions proto.
std::unique_ptr<ImageSegmenterGraphOptionsProto>
ConvertImageSegmenterOptionsToProto(ImageSegmenterOptions* options) {
  auto options_proto = std::make_unique<ImageSegmenterGraphOptionsProto>();
  auto base_options_proto = std::make_unique<tasks::core::proto::BaseOptions>(
      tasks::core::ConvertBaseOptionsToProto(&(options->base_options)));
  options_proto->mutable_base_options()->Swap(base_options_proto.get());
  options_proto->mutable_base_options()->set_use_stream_mode(
      options->running_mode != core::RunningMode::IMAGE);
  options_proto->set_display_names_locale(options->display_names_locale);
  return options_proto;
}

absl::StatusOr<std::vector<std::string>> GetLabelsFromGraphConfig(
    const CalculatorGraphConfig& graph_config) {
  bool found_tensor_to_segmentation_calculator = false;
  std::vector<std::string> labels;
  for (const auto& node : graph_config.node()) {
    if (node.calculator() ==
        "mediapipe.tasks.TensorsToSegmentationCalculator") {
      if (!found_tensor_to_segmentation_calculator) {
        found_tensor_to_segmentation_calculator = true;
      } else {
        return absl::Status(CreateStatusWithPayload(
            absl::StatusCode::kFailedPrecondition,
            "The graph has more than one "
            "mediapipe.tasks.TensorsToSegmentationCalculator."));
      }
      TensorsToSegmentationCalculatorOptions options =
          node.options().GetExtension(
              TensorsToSegmentationCalculatorOptions::ext);
      if (!options.label_items().empty()) {
        for (int i = 0; i < options.label_items_size(); ++i) {
          if (!options.label_items().contains(i)) {
            return absl::Status(CreateStatusWithPayload(
                absl::StatusCode::kFailedPrecondition,
                absl::StrFormat("The lablemap have no expected key: %d.", i)));
          }
          labels.push_back(options.label_items().at(i).name());
        }
      }
    }
  }
  return labels;
}

}  // namespace

absl::StatusOr<std::unique_ptr<ImageSegmenter>> ImageSegmenter::Create(
    std::unique_ptr<ImageSegmenterOptions> options) {
  if (!options->output_confidence_masks && !options->output_category_mask) {
    return absl::InvalidArgumentError(
        "At least one of `output_confidence_masks` and `output_category_mask` "
        "must be set.");
  }
  auto options_proto = ConvertImageSegmenterOptionsToProto(options.get());
  tasks::core::PacketsCallback packets_callback = nullptr;
  if (options->result_callback) {
    auto result_callback = options->result_callback;
    bool output_category_mask = options->output_category_mask;
    bool output_confidence_masks = options->output_confidence_masks;
    packets_callback =
        [=](absl::StatusOr<tasks::core::PacketMap> status_or_packets) {
          if (!status_or_packets.ok()) {
            Image image;
            result_callback(status_or_packets.status(), image,
                            Timestamp::Unset().Value());
            return;
          }
          if (status_or_packets.value()[kImageOutStreamName].IsEmpty()) {
            return;
          }
          std::optional<std::vector<Image>> confidence_masks;
          if (output_confidence_masks) {
            confidence_masks =
                status_or_packets.value()[kConfidenceMasksStreamName]
                    .Get<std::vector<Image>>();
          }
          std::optional<Image> category_mask;
          if (output_category_mask) {
            category_mask =
                status_or_packets.value()[kCategoryMaskStreamName].Get<Image>();
          }
          const std::vector<float>& quality_scores =
              status_or_packets.value()[kQualityScoresStreamName]
                  .Get<std::vector<float>>();
          Packet image_packet = status_or_packets.value()[kImageOutStreamName];
          result_callback(
              {{confidence_masks, category_mask, quality_scores}},
              image_packet.Get<Image>(),
              image_packet.Timestamp().Value() / kMicroSecondsPerMilliSecond);
        };
  }
  auto image_segmenter =
      core::VisionTaskApiFactory::Create<ImageSegmenter,
                                         ImageSegmenterGraphOptionsProto>(
          CreateGraphConfig(
              std::move(options_proto), options->output_confidence_masks,
              options->output_category_mask,
              options->running_mode == core::RunningMode::LIVE_STREAM),
          std::move(options->base_options.op_resolver), options->running_mode,
          std::move(packets_callback),
          /*disable_default_service=*/
          options->base_options.disable_default_service);
  if (!image_segmenter.ok()) {
    return image_segmenter.status();
  }
  image_segmenter.value()->output_confidence_masks_ =
      options->output_confidence_masks;
  image_segmenter.value()->output_category_mask_ =
      options->output_category_mask;
  MP_ASSIGN_OR_RETURN(
      (*image_segmenter)->labels_,
      GetLabelsFromGraphConfig((*image_segmenter)->runner_->GetGraphConfig()));
  return image_segmenter;
}

absl::StatusOr<ImageSegmenterResult> ImageSegmenter::Segment(
    mediapipe::Image image,
    std::optional<core::ImageProcessingOptions> image_processing_options) {
  return Segment(image, {
                            /*output_width=*/image.width(),
                            /*output_height=*/image.height(),
                            std::move(image_processing_options),
                        });
}

absl::StatusOr<ImageSegmenterResult> ImageSegmenter::Segment(
    mediapipe::Image image, SegmentationOptions segmentation_options) {
  MP_RETURN_IF_ERROR(ValidateSegmentationOptions(segmentation_options));
  if (image.UsesGpu()) {
    return CreateStatusWithPayload(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("GPU input images are currently not supported."),
        MediaPipeTasksStatus::kRunnerUnexpectedInputError);
  }
  MP_ASSIGN_OR_RETURN(NormalizedRect norm_rect,
                      ConvertToNormalizedRect(
                          segmentation_options.image_processing_options, image,
                          /*roi_allowed=*/false));
  MP_ASSIGN_OR_RETURN(
      auto output_packets,
      ProcessImageData(
          {{kImageInStreamName, mediapipe::MakePacket<Image>(std::move(image))},
           {kNormRectStreamName,
            MakePacket<NormalizedRect>(std::move(norm_rect))},
           {kOutputSizeStreamName,
            MakePacket<std::pair<int, int>>(
                std::make_pair(segmentation_options.output_width,
                               segmentation_options.output_height))}}));
  std::optional<std::vector<Image>> confidence_masks;
  if (output_confidence_masks_) {
    confidence_masks =
        output_packets[kConfidenceMasksStreamName].Get<std::vector<Image>>();
  }
  std::optional<Image> category_mask;
  if (output_category_mask_) {
    category_mask = output_packets[kCategoryMaskStreamName].Get<Image>();
  }
  const std::vector<float>& quality_scores =
      output_packets[kQualityScoresStreamName].Get<std::vector<float>>();
  return {{confidence_masks, category_mask, quality_scores}};
}

absl::StatusOr<ImageSegmenterResult> ImageSegmenter::SegmentForVideo(
    mediapipe::Image image, int64_t timestamp_ms,
    std::optional<core::ImageProcessingOptions> image_processing_options) {
  return SegmentForVideo(image, timestamp_ms,
                         {
                             /*output_width=*/image.width(),
                             /*output_height=*/image.height(),
                             std::move(image_processing_options),
                         });
}

absl::StatusOr<ImageSegmenterResult> ImageSegmenter::SegmentForVideo(
    mediapipe::Image image, int64_t timestamp_ms,
    SegmentationOptions segmentation_options) {
  MP_RETURN_IF_ERROR(ValidateSegmentationOptions(segmentation_options));
  if (image.UsesGpu()) {
    return CreateStatusWithPayload(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("GPU input images are currently not supported."),
        MediaPipeTasksStatus::kRunnerUnexpectedInputError);
  }
  MP_ASSIGN_OR_RETURN(NormalizedRect norm_rect,
                      ConvertToNormalizedRect(
                          segmentation_options.image_processing_options, image,
                          /*roi_allowed=*/false));
  MP_ASSIGN_OR_RETURN(
      auto output_packets,
      ProcessVideoData(
          {{kImageInStreamName,
            MakePacket<Image>(std::move(image))
                .At(Timestamp(timestamp_ms * kMicroSecondsPerMilliSecond))},
           {kNormRectStreamName,
            MakePacket<NormalizedRect>(std::move(norm_rect))
                .At(Timestamp(timestamp_ms * kMicroSecondsPerMilliSecond))},
           {kOutputSizeStreamName,
            MakePacket<std::pair<int, int>>(
                std::make_pair(segmentation_options.output_width,
                               segmentation_options.output_height))
                .At(Timestamp(timestamp_ms * kMicroSecondsPerMilliSecond))}}));
  std::optional<std::vector<Image>> confidence_masks;
  if (output_confidence_masks_) {
    confidence_masks =
        output_packets[kConfidenceMasksStreamName].Get<std::vector<Image>>();
  }
  std::optional<Image> category_mask;
  if (output_category_mask_) {
    category_mask = output_packets[kCategoryMaskStreamName].Get<Image>();
  }
  const std::vector<float>& quality_scores =
      output_packets[kQualityScoresStreamName].Get<std::vector<float>>();
  return {{confidence_masks, category_mask, quality_scores}};
}

absl::Status ImageSegmenter::SegmentAsync(
    Image image, int64_t timestamp_ms,
    std::optional<core::ImageProcessingOptions> image_processing_options) {
  return SegmentAsync(image, timestamp_ms,
                      {
                          /*output_width=*/image.width(),
                          /*output_height=*/image.height(),
                          std::move(image_processing_options),
                      });
}

absl::Status ImageSegmenter::SegmentAsync(
    Image image, int64_t timestamp_ms,
    SegmentationOptions segmentation_options) {
  MP_RETURN_IF_ERROR(ValidateSegmentationOptions(segmentation_options));
  if (image.UsesGpu()) {
    return CreateStatusWithPayload(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("GPU input images are currently not supported."),
        MediaPipeTasksStatus::kRunnerUnexpectedInputError);
  }
  MP_ASSIGN_OR_RETURN(NormalizedRect norm_rect,
                      ConvertToNormalizedRect(
                          segmentation_options.image_processing_options, image,
                          /*roi_allowed=*/false));
  return SendLiveStreamData(
      {{kImageInStreamName,
        MakePacket<Image>(std::move(image))
            .At(Timestamp(timestamp_ms * kMicroSecondsPerMilliSecond))},
       {kNormRectStreamName,
        MakePacket<NormalizedRect>(std::move(norm_rect))
            .At(Timestamp(timestamp_ms * kMicroSecondsPerMilliSecond))},
       {kOutputSizeStreamName,
        MakePacket<std::pair<int, int>>(
            std::make_pair(segmentation_options.output_width,
                           segmentation_options.output_height))
            .At(Timestamp(timestamp_ms * kMicroSecondsPerMilliSecond))}});
}

}  // namespace image_segmenter
}  // namespace vision
}  // namespace tasks
}  // namespace mediapipe

File: mediapipe/tasks/web/vision/image_segmenter/image_segmenter.ts
/**
 * Copyright 2022 The MediaPipe Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {CalculatorGraphConfig} from '../../../../framework/calculator_pb';
import {CalculatorOptions} from '../../../../framework/calculator_options_pb';
import {BaseOptions as BaseOptionsProto} from '../../../../tasks/cc/core/proto/base_options_pb';
import {TensorsToSegmentationCalculatorOptions} from '../../../../tasks/cc/vision/image_segmenter/calculators/tensors_to_segmentation_calculator_pb';
import {ImageSegmenterGraphOptions as ImageSegmenterGraphOptionsProto} from '../../../../tasks/cc/vision/image_segmenter/proto/image_segmenter_graph_options_pb';
import {SegmenterOptions as SegmenterOptionsProto} from '../../../../tasks/cc/vision/image_segmenter/proto/segmenter_options_pb';
import {WasmFileset} from '../../../../tasks/web/core/wasm_fileset';
import {ImageProcessingOptions} from '../../../../tasks/web/vision/core/image_processing_options';
import {MPMask} from '../../../../tasks/web/vision/core/mask';
import {
  VisionGraphRunner,
  VisionTaskRunner,
} from '../../../../tasks/web/vision/core/vision_task_runner';
import {LabelMapItem} from '../../../../util/label_map_pb';
import {
  ImageSource,
  WasmModule,
} from '../../../../web/graph_runner/graph_runner';
// Placeholder for internal dependency on trusted resource url

import {ImageSegmenterOptions} from './image_segmenter_options';
import {ImageSegmenterResult} from './image_segmenter_result';

export * from './image_segmenter_options';
export * from './image_segmenter_result';
export {type ImageSource}; // Used in the public API

const IMAGE_STREAM = 'image_in';
const NORM_RECT_STREAM = 'norm_rect';
const CONFIDENCE_MASKS_STREAM = 'confidence_masks';
const CATEGORY_MASK_STREAM = 'category_mask';
const QUALITY_SCORES_STREAM = 'quality_scores';
const IMAGE_SEGMENTER_GRAPH =
  'mediapipe.tasks.vision.image_segmenter.ImageSegmenterGraph';
const TENSORS_TO_SEGMENTATION_CALCULATOR_NAME =
  'mediapipe.tasks.TensorsToSegmentationCalculator';
const DEFAULT_OUTPUT_CATEGORY_MASK = false;
const DEFAULT_OUTPUT_CONFIDENCE_MASKS = true;

// The OSS JS API does not support the builder pattern.
// tslint:disable:jspb-use-builder-pattern

/**
 * A callback that receives the computed masks from the image segmenter. The
 * returned data is only valid for the duration of the callback. If
 * asynchronous processing is needed, all data needs to be copied before the
 * callback returns.
 */
export type ImageSegmenterCallback = (result: ImageSegmenterResult) => void;

/** Performs image segmentation on images. */
export class ImageSegmenter extends VisionTaskRunner {
  private categoryMask?: MPMask;
  private confidenceMasks?: MPMask[];
  private qualityScores?: number[];
  private labels: string[] = [];
  private userCallback?: ImageSegmenterCallback;
  private outputCategoryMask = DEFAULT_OUTPUT_CATEGORY_MASK;
  private outputConfidenceMasks = DEFAULT_OUTPUT_CONFIDENCE_MASKS;
  private readonly options: ImageSegmenterGraphOptionsProto;
  private readonly segmenterOptions: SegmenterOptionsProto;

  /**
   * Initializes the Wasm runtime and creates a new image segmenter from the
   * provided options.
   * @export
   * @param wasmFileset A configuration object that provides the location of
   *     the Wasm binary and its loader.
   * @param imageSegmenterOptions The options for the Image Segmenter. Note
   *     that either a path to the model asset or a model buffer needs to be
   *     provided (via `baseOptions`).
   */
  static createFromOptions(
    wasmFileset: WasmFileset,
    imageSegmenterOptions: ImageSegmenterOptions,
  ): Promise<ImageSegmenter> {
    return VisionTaskRunner.createVisionInstance(
      ImageSegmenter,
      wasmFileset,
      imageSegmenterOptions,
    );
  }

  /**
   * Initializes the Wasm runtime and creates a new image segmenter based on
   * the provided model asset buffer.
   * @export
   * @param wasmFileset A configuration object that provides the location of
   *     the Wasm binary and its loader.
   * @param modelAssetBuffer An array or a stream containing a binary
   *    representation of the model.
   */
  static createFromModelBuffer(
    wasmFileset: WasmFileset,
    modelAssetBuffer: Uint8Array | ReadableStreamDefaultReader,
  ): Promise<ImageSegmenter> {
    return VisionTaskRunner.createVisionInstance(ImageSegmenter, wasmFileset, {
      baseOptions: {modelAssetBuffer},
    });
  }

  /**
   * Initializes the Wasm runtime and creates a new image segmenter based on
   * the path to the model asset.
   * @export
   * @param wasmFileset A configuration object that provides the location of
   *     the Wasm binary and its loader.
   * @param modelAssetPath The path to the model asset.
   */
  static createFromModelPath(
    wasmFileset: WasmFileset,
    modelAssetPath: string,
  ): Promise<ImageSegmenter> {
    return VisionTaskRunner.createVisionInstance(ImageSegmenter, wasmFileset, {
      baseOptions: {modelAssetPath},
    });
  }

  /** @hideconstructor */
  constructor(
    wasmModule: WasmModule,
    glCanvas?: HTMLCanvasElement | OffscreenCanvas | null,
  ) {
    super(
      new VisionGraphRunner(wasmModule, glCanvas),
      IMAGE_STREAM,
      NORM_RECT_STREAM,
      /* roiAllowed= */ false,
    );
    this.options = new ImageSegmenterGraphOptionsProto();
    this.segmenterOptions = new SegmenterOptionsProto();
    this.options.setSegmenterOptions(this.segmenterOptions);
    this.options.setBaseOptions(new BaseOptionsProto());
  }

  protected override get baseOptions(): BaseOptionsProto {
    return this.options.getBaseOptions()!;
  }

  protected override set baseOptions(proto: BaseOptionsProto) {
    this.options.setBaseOptions(proto);
  }

  /**
   * Sets new options for the image segmenter.
   *
   * Calling `setOptions()` with a subset of options only affects those
   * options. You can reset an option back to its default value by
   * explicitly setting it to `undefined`.
   *
   * @export
   * @param options The options for the image segmenter.
   */
  override setOptions(options: ImageSegmenterOptions): Promise<void> {
    // Note that we have to support both JSPB and ProtobufJS, hence we
    // have to expliclity clear the values instead of setting them to
    // `undefined`.
    if (options.displayNamesLocale !== undefined) {
      this.options.setDisplayNamesLocale(options.displayNamesLocale);
    } else if ('displayNamesLocale' in options) {
      // Check for undefined
      this.options.clearDisplayNamesLocale();
    }

    if ('outputCategoryMask' in options) {
      this.outputCategoryMask =
        options.outputCategoryMask ?? DEFAULT_OUTPUT_CATEGORY_MASK;
    }

    if ('outputConfidenceMasks' in options) {
      this.outputConfidenceMasks =
        options.outputConfidenceMasks ?? DEFAULT_OUTPUT_CONFIDENCE_MASKS;
    }

    return super.applyOptions(options);
  }

  protected override onGraphRefreshed(): void {
    this.populateLabels();
  }

  /**
   * Populate the labelMap in TensorsToSegmentationCalculator to labels field.
   * @throws Exception if there is an error during finding
   *     TensorsToSegmentationCalculator.
   */
  private populateLabels(): void {
    const graphConfig = this.getCalculatorGraphConfig();
    const tensorsToSegmentationCalculators = graphConfig
      .getNodeList()
      .filter((n: CalculatorGraphConfig.Node) =>
        n.getName().includes(TENSORS_TO_SEGMENTATION_CALCULATOR_NAME),
      );

    this.labels = [];
    if (tensorsToSegmentationCalculators.length > 1) {
      throw new Error(
        `The graph has more than one ${TENSORS_TO_SEGMENTATION_CALCULATOR_NAME}.`,
      );
    } else if (tensorsToSegmentationCalculators.length === 1) {
      const labelItems =
        tensorsToSegmentationCalculators[0]
          .getOptions()
          ?.getExtension(TensorsToSegmentationCalculatorOptions.ext)
          ?.getLabelItemsMap() ?? new Map<string, LabelMapItem>();
      labelItems.forEach((value, index) => {
        // tslint:disable-next-line:no-unnecessary-type-assertion
        this.labels[Number(index)] = value.getName()!;
      });
    }
  }

  /**
   * Performs image segmentation on the provided single image and invokes the
   * callback with the response. The method returns synchronously once the
   * callback returns. Only use this method when the ImageSegmenter is
   * created with running mode `image`.
   *
   * @param image An image to process.
   * @param callback The callback that is invoked with the segmented masks. The
   *    lifetime of the returned data is only guaranteed for the duration of the
   *    callback.
   */
  segment(image: ImageSource, callback: ImageSegmenterCallback): void;
  /**
   * Performs image segmentation on the provided single image and invokes the
   * callback with the response. The method returns synchronously once the
   * callback returns. Only use this method when the ImageSegmenter is
   * created with running mode `image`.
   *
   * @param image An image to process.
   * @param imageProcessingOptions the `ImageProcessingOptions` specifying how
   *    to process the input image before running inference.
   * @param callback The callback that is invoked with the segmented masks. The
   *    lifetime of the returned data is only guaranteed for the duration of the
   *    callback.
   */
  segment(
    image: ImageSource,
    imageProcessingOptions: ImageProcessingOptions,
    callback: ImageSegmenterCallback,
  ): void;
  /**
   * Performs image segmentation on the provided single image and returns the
   * segmentation result. This method creates a copy of the resulting masks and
   * should not be used in high-throughput applications. Only use this method
   * when the ImageSegmenter is created with running mode `image`.
   *
   * @param image An image to process.
   * @return The segmentation result. The data is copied to avoid lifetime
   *     issues.
   */
  segment(image: ImageSource): ImageSegmenterResult;
  /**
   * Performs image segmentation on the provided single image and returns the
   * segmentation result. This method creates a copy of the resulting masks and
   * should not be used in high-v applications. Only use this method when
   * the ImageSegmenter is created with running mode `image`.
   *
   * @param image An image to process.
   * @param imageProcessingOptions the `ImageProcessingOptions` specifying how
   *    to process the input image before running inference.
   * @return The segmentation result. The data is copied to avoid lifetime
   *     issues.
   */
  segment(
    image: ImageSource,
    imageProcessingOptions: ImageProcessingOptions,
  ): ImageSegmenterResult;
  /** @export */
  segment(
    image: ImageSource,
    imageProcessingOptionsOrCallback?:
      | ImageProcessingOptions
      | ImageSegmenterCallback,
    callback?: ImageSegmenterCallback,
  ): ImageSegmenterResult | void {
    const imageProcessingOptions =
      typeof imageProcessingOptionsOrCallback !== 'function'
        ? imageProcessingOptionsOrCallback
        : {};

    this.userCallback =
      typeof imageProcessingOptionsOrCallback === 'function'
        ? imageProcessingOptionsOrCallback
        : callback;

    this.reset();
    this.processImageData(image, imageProcessingOptions);
    return this.processResults();
  }

  /**
   * Performs image segmentation on the provided video frame and invokes the
   * callback with the response. The method returns synchronously once the
   * callback returns. Only use this method when the ImageSegmenter is
   * created with running mode `video`.
   *
   * @param videoFrame A video frame to process.
   * @param timestamp The timestamp of the current frame, in ms.
   * @param callback The callback that is invoked with the segmented masks. The
   *    lifetime of the returned data is only guaranteed for the duration of the
   *    callback.
   */
  segmentForVideo(
    videoFrame: ImageSource,
    timestamp: number,
    callback: ImageSegmenterCallback,
  ): void;
  /**
   * Performs image segmentation on the provided video frame and invokes the
   * callback with the response. The method returns synchronously once the
   * callback returns. Only use this method when the ImageSegmenter is
   * created with running mode `video`.
   *
   * @param videoFrame A video frame to process.
   * @param timestamp The timestamp of the current frame, in ms.
   * @param imageProcessingOptions the `ImageProcessingOptions` specifying how
   *    to process the input frame before running inference.
   * @param callback The callback that is invoked with the segmented masks. The
   *    lifetime of the returned data is only guaranteed for the duration of the
   *    callback.
   */
  segmentForVideo(
    videoFrame: ImageSource,
    timestamp: number,
    imageProcessingOptions: ImageProcessingOptions,
    callback: ImageSegmenterCallback,
  ): void;
  /**
   * Performs image segmentation on the provided video frame and returns the
   * segmentation result. This method creates a copy of the resulting masks and
   * should not be used in high-throughput applications. Only use this method
   * when the ImageSegmenter is created with running mode `video`.
   *
   * @param videoFrame A video frame to process.
   * @return The segmentation result. The data is copied to avoid lifetime
   *     issues.
   */
  segmentForVideo(
    videoFrame: ImageSource,
    timestamp: number,
  ): ImageSegmenterResult;
  /**
   * Performs image segmentation on the provided video frame and returns the
   * segmentation result. This method creates a copy of the resulting masks and
   * should not be used in high-v applications. Only use this method when
   * the ImageSegmenter is created with running mode `video`.
   *
   * @param videoFrame A video frame to process.
   * @param timestamp The timestamp of the current frame, in ms.
   * @param imageProcessingOptions the `ImageProcessingOptions` specifying how
   *    to process the input frame before running inference.
   * @return The segmentation result. The data is copied to avoid lifetime
   *     issues.
   */
  segmentForVideo(
    videoFrame: ImageSource,
    timestamp: number,
    imageProcessingOptions: ImageProcessingOptions,
  ): ImageSegmenterResult;
  /** @export */
  segmentForVideo(
    videoFrame: ImageSource,
    timestamp: number,
    imageProcessingOptionsOrCallback?:
      | ImageProcessingOptions
      | ImageSegmenterCallback,
    callback?: ImageSegmenterCallback,
  ): ImageSegmenterResult | void {
    const imageProcessingOptions =
      typeof imageProcessingOptionsOrCallback !== 'function'
        ? imageProcessingOptionsOrCallback
        : {};
    this.userCallback =
      typeof imageProcessingOptionsOrCallback === 'function'
        ? imageProcessingOptionsOrCallback
        : callback;

    this.reset();
    this.processVideoData(videoFrame, imageProcessingOptions, timestamp);
    return this.processResults();
  }

  /**
   * Get the category label list of the ImageSegmenter can recognize. For
   * `CATEGORY_MASK` type, the index in the category mask corresponds to the
   * category in the label list. For `CONFIDENCE_MASK` type, the output mask
   * list at index corresponds to the category in the label list.
   *
   * If there is no labelmap provided in the model file, empty label array is
   * returned.
   *
   * @export
   * @return The labels used by the current model.
   */
  getLabels(): string[] {
    return this.labels;
  }

  private reset(): void {
    this.categoryMask = undefined;
    this.confidenceMasks = undefined;
    this.qualityScores = undefined;
  }

  private processResults(): ImageSegmenterResult | void {
    try {
      const result = new ImageSegmenterResult(
        this.confidenceMasks,
        this.categoryMask,
        this.qualityScores,
      );
      if (this.userCallback) {
        this.userCallback(result);
      } else {
        return result;
      }
    } finally {
      // Free the image memory, now that we've kept all streams alive long
      // enough to be returned in our callbacks.
      this.freeKeepaliveStreams();
    }
  }

  /** Updates the MediaPipe graph configuration. */
  protected override refreshGraph(): void {
    const graphConfig = new CalculatorGraphConfig();
    graphConfig.addInputStream(IMAGE_STREAM);
    graphConfig.addInputStream(NORM_RECT_STREAM);

    const calculatorOptions = new CalculatorOptions();
    calculatorOptions.setExtension(
      ImageSegmenterGraphOptionsProto.ext,
      this.options,
    );

    const segmenterNode = new CalculatorGraphConfig.Node();
    segmenterNode.setCalculator(IMAGE_SEGMENTER_GRAPH);
    segmenterNode.addInputStream('IMAGE:' + IMAGE_STREAM);
    segmenterNode.addInputStream('NORM_RECT:' + NORM_RECT_STREAM);
    segmenterNode.setOptions(calculatorOptions);

    graphConfig.addNode(segmenterNode);
    this.addKeepaliveNode(graphConfig);

    if (this.outputConfidenceMasks) {
      graphConfig.addOutputStream(CONFIDENCE_MASKS_STREAM);
      segmenterNode.addOutputStream(
        'CONFIDENCE_MASKS:' + CONFIDENCE_MASKS_STREAM,
      );
      this.keepStreamAlive(CONFIDENCE_MASKS_STREAM);

      this.graphRunner.attachImageVectorListener(
        CONFIDENCE_MASKS_STREAM,
        (masks, timestamp) => {
          this.confidenceMasks = masks.map((wasmImage) =>
            this.convertToMPMask(
              wasmImage,
              /* interpolateValues= */ true,
              /* shouldCopyData= */ !this.userCallback,
            ),
          );
          this.setLatestOutputTimestamp(timestamp);
        },
      );
      this.graphRunner.attachEmptyPacketListener(
        CONFIDENCE_MASKS_STREAM,
        (timestamp) => {
          this.confidenceMasks = [];
          this.setLatestOutputTimestamp(timestamp);
        },
      );
    }

    if (this.outputCategoryMask) {
      graphConfig.addOutputStream(CATEGORY_MASK_STREAM);
      segmenterNode.addOutputStream('CATEGORY_MASK:' + CATEGORY_MASK_STREAM);
      this.keepStreamAlive(CATEGORY_MASK_STREAM);

      this.graphRunner.attachImageListener(
        CATEGORY_MASK_STREAM,
        (mask, timestamp) => {
          this.categoryMask = this.convertToMPMask(
            mask,
            /* interpolateValues= */ false,
            /* shouldCopyData= */ !this.userCallback,
          );
          this.setLatestOutputTimestamp(timestamp);
        },
      );
      this.graphRunner.attachEmptyPacketListener(
        CATEGORY_MASK_STREAM,
        (timestamp) => {
          this.categoryMask = undefined;
          this.setLatestOutputTimestamp(timestamp);
        },
      );
    }

    graphConfig.addOutputStream(QUALITY_SCORES_STREAM);
    segmenterNode.addOutputStream('QUALITY_SCORES:' + QUALITY_SCORES_STREAM);

    this.graphRunner.attachFloatVectorListener(
      QUALITY_SCORES_STREAM,
      (scores, timestamp) => {
        this.qualityScores = scores;
        this.setLatestOutputTimestamp(timestamp);
      },
    );
    this.graphRunner.attachEmptyPacketListener(
      QUALITY_SCORES_STREAM,
      (timestamp) => {
        this.categoryMask = undefined;
        this.setLatestOutputTimestamp(timestamp);
      },
    );

    const binaryGraph = graphConfig.serializeBinary();
    this.setGraph(new Uint8Array(binaryGraph), /* isBinary= */ true);
  }
}


